logging/logger.go

```diff
@@ -0,0 +1,293 @@
+// library for Go application code or modules. Public (to the library) methods that take a logger as
+// input should accept this type and upconvert to a Logger via a call to `FromZapCompatible`.
+type ZapCompatibleLogger interface {
+	Desugar() *zap.Logger
+	Level() zapcore.Level
+	Named(name string) *zap.SugaredLogger
+	Sync() error
+	With(args ...interface{}) *zap.SugaredLogger
+	WithOptions(opts ...zap.Option) *zap.SugaredLogger
+
+	Debug(args ...interface{})
+	Debugf(template string, args ...interface{})
+	Debugln(args ...interface{})
+	Debugw(msg string, keysAndValues ...interface{})
+
+	Info(args ...interface{})
+	Infof(template string, args ...interface{})
+	Infoln(args ...interface{})
+	Infow(msg string, keysAndValues ...interface{})
+
+	Warn(args ...interface{})
+	Warnf(template string, args ...interface{})
+	Warnln(args ...interface{})
+	Warnw(msg string, keysAndValues ...interface{})
+
+	Error(args ...interface{})
+	Errorf(template string, args ...interface{})
+	Errorln(args ...interface{})
+	Errorw(msg string, keysAndValues ...interface{})
+
+	Fatal(args ...interface{})
+	Fatalf(template string, args ...interface{})
+	Fatalln(args ...interface{})
+	Fatalw(msg string, keysAndValues ...interface{})
+
+	Panic(args ...interface{})
+	Panicf(template string, args ...interface{})
+	Panicln(args ...interface{})
+	Panicw(msg string, keysAndValues ...interface{})
+
+	DPanic(args ...interface{})
+	DPanicf(template string, args ...interface{})
+	DPanicln(args ...interface{})
+	DPanicw(msg string, keysAndValues ...interface{})
+}
+
+// ZLogger type for logging to. Wraps a zap logger and adds the `AsZap` method to satisfy the
+// `Logger` interface.
+type zLogger struct {
+	*zap.SugaredLogger
+}
+
+// FromZapCompatible upconverts a ZapCompatibleLogger to a logging.Logger. If the argument already
+// satisfies logging.Logger, no changes will be made. A nil input returns a nil logger. An input of
+// unknown type will create a new logger that's not associated with the input.
+func FromZapCompatible(logger ZapCompatibleLogger) Logger {
+	if logger == nil {
+		return nil
+	}
+
+	switch l := logger.(type) {
+	case *zap.SugaredLogger:
+		// golog.Logger is a type alias for *zap.SugaredLogger and is captured by this.
+		return zLogger{l}
+	case Logger:
+		return l
+	default:
+		logger.Warnf("Unknown logger type, creating a new Viam Logger. Unknown type: %T", logger)
+		return newDefaultLogger()
+	}
+}
+
+// AsZap converts the logger to a zap logger.
+func (logger zLogger) AsZap() *zap.SugaredLogger {
+	return logger.SugaredLogger
+}
+
+var (
+	globalMu     sync.RWMutex
+	globalLogger = newDefaultLogger()
+)
+
+func newDefaultLogger() Logger {
+	return zLogger{zap.Must(NewDebugLoggerConfig().Build()).Sugar()}
+}
+
+// ReplaceGloabl replaces the global loggers and returns a function to reset
+// the loggers to the previous state.
+func ReplaceGloabl(logger Logger) func() {
```

Comment:
[nit] This misspelling has always bothered me but I assume we have to maintain the signature for backward compatibility? Since we may break Go modules here already, can we rename this?

Reply:
lol yes it bothers me too. And this global logger business is actually a bit worse than just a mispelling. Goutils has a [global logger variable](https://github.com/viamrobotics/goutils/blob/main/logger.go#L8) that we use in a [couple](https://github.com/viamrobotics/rdk/blob/39e81fc74f3b302228ea1093cb389f7fa80184d6/components/encoder/incremental/incremental_encoder.go#L219) [spots](https://github.com/viamrobotics/rdk/blob/39e81fc74f3b302228ea1093cb389f7fa80184d6/pointcloud/icp.go#L116).

That logger is actually the "[original](https://github.com/edaniels/golog/blob/28954395e8d0e7ae7f0046e4c845d553b94e4805/logger.go#L41)" global logger. Meaning it doesn't track changes via `ReplaceGloabl`.

If you don't mind I'd prefer to do some of this other small cleanup in separate patches/PRs. Though I admit there's a risk that we can "break" compatibility twice if those patches land in adjacent releases. But I also don't think there's a real risk of modules calling this today.

Reply:
> If you don't mind I'd prefer to do some of this other small cleanup in separate patches/PRs. Though I admit there's a risk that we can "break" compatibility twice if those patches land in adjacent releases. But I also don't think there's a real risk of modules calling this today.

Agree and SGTM.